src/main/java/org/folio/edge/sip2/MainVerticle.java

```diff
@@ -103,148 +108,190 @@ public void start(Promise<Void> startFuture) {
-          log.error("No tenant configured for address: {}  message ignored.", clientAddress);
-          return;
-        }
-
-        final String messageString = buffer.getString(0, buffer.length(), sessionData.getCharset());
-
-        log.debug("Received message: {}", messageString);
-
-        Command command = UNKNOWN;
-
-        try {
-          final Parser parser = Parser.builder()
-              .delimiter(sessionData.getFieldDelimiter())
-              .charset(Charset.forName(sessionData.getCharset()))
-              .errorDetectionEnabled(sessionData.isErrorDetectionEnabled())
-              .timezone(sessionData.getTimeZone())
-              .build();
-
-          //parsing
-          final Message<Object> message = parser.parseMessage(messageString);
+    var portList = determinePorts();
+
+    if (!portList.isEmpty()) {
+      AtomicInteger remainingServers = new AtomicInteger(portList.size());
+
+      for (int i = 0; i < portList.size(); i++) {
+        int port = portList.get(i);
+        NetServerOptions options = new NetServerOptions(
+            config().getJsonObject("netServerOptions", new JsonObject()))
+            .setPort(port);
+
+        NetServer server = vertx.createNetServer(options);
+        servers.add(server);
+
+        final Metrics metrics = Metrics.getMetrics(port);
+        metricsMap.putIfAbsent(port, metrics);
+
+        server.connectHandler(socket -> {
+
+          String clientAddress = socket.remoteAddress().host();
+
+          ThreadContext.put(IPADDRESS, clientAddress);
+
+          JsonObject tenantConfig = TenantUtils.lookupTenantConfigForIPaddress(multiTenantConfig,
+              clientAddress, port);
+
+          final SessionData sessionData = SessionData.createSession(
+              tenantConfig.getString("tenant"),
+              tenantConfig.getString("fieldDelimiter", "|").charAt(0),
+              tenantConfig.getBoolean("errorDetectionEnabled", FALSE),
+              tenantConfig.getString("charset", "IBM850"));
+          final String messageDelimiter = tenantConfig.getString("messageDelimiter", "\r");
+
+          socket.handler(RecordParser.newDelimited(messageDelimiter, buffer -> {
+            final Timer.Sample sample = metrics.sample();
+
+            if (Objects.isNull(sessionData.getTenant())) {
+              log.error("No tenant configured for address: {}  message ignored.", clientAddress);
+              return;
+            }
+
+            final String messageString = buffer.getString(0, buffer.length(),
+                sessionData.getCharset());
+
+            Command command = UNKNOWN;
+
+            try {
+              final Parser parser = Parser.builder()
+                  .delimiter(sessionData.getFieldDelimiter())
+                  .charset(Charset.forName(sessionData.getCharset()))
+                  .errorDetectionEnabled(sessionData.isErrorDetectionEnabled())
+                  .timezone(sessionData.getTimeZone())
+                  .build();
```

Comment:
Maybe put this in a method called `getParser`. Doesn't reduce cognitive complexity score, but I think makes it more readable.

Reply:
done